client/block.c

```diff
@@ -559,6 +640,50 @@ static int add_block_nolock(struct xdag_block *newBlock, xdag_time_t limit)
 		}
 	}
 
+#if CACHE == 1
+	if(g_xdag_extstats.cache_usage >= g_xdag_extstats.cache_size){
+		if (g_xdag_extstats.cache_hitrate<0.94 && g_xdag_extstats.cache_size*2 <= CACHE_MAX_SIZE){
+			if(!g_xdag_extstats.cache_size && CACHE_MAX_SIZE){
+				g_xdag_extstats.cache_size++;
+			}
+			else{
+				g_xdag_extstats.cache_size = g_xdag_extstats.cache_size*2;
+			}
+		}
+		else if(g_xdag_extstats.cache_hitrate>0.98 && !cache_miss && g_xdag_extstats.cache_size){
+			g_xdag_extstats.cache_size--;
+		}
+		for(int l=g_xdag_extstats.cache_usage;l>g_xdag_extstats.cache_size;l--){
+			if(cache_first != NULL){
+				struct cache_block* to_free = cache_first;
```

Comment:
Is it really necessary to have both linked list and tree for cache? Am I right the only reason is to always remove the first element. Like FIFO queue (First In First Out)?